### PR TITLE
Improve convert-to-file-scoped-namespace trivia handling.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/63542
Fixes https://github.com/dotnet/roslyn/issues/63792

### CODE REVIEW

File: src/Analyzers/CSharp/CodeFixes/ConvertNamespace/ConvertNamespaceTransform.cs [hunk 1]
```diff
@@ -186,53 +186,56 @@ public static (SourceText text, TextSpan semicolonSpan) ConvertNamespaceDeclarat
...
         {
-            var semiColon = SyntaxFactory.Token(SyntaxKind.SemicolonToken).WithoutTrivia();
+            // If the open-brace token has any special trivia, then move them to after the semicolon.
+            var semiColon = SyntaxFactory.Token(SyntaxKind.SemicolonToken)
+                .WithoutTrivia()
+                .WithTrailingTrivia(namespaceDeclaration.Name.GetTrailingTrivia())
+                .WithAppendedTrailingTrivia(namespaceDeclaration.OpenBraceToken.LeadingTrivia);
+
+            if (!namespaceDeclaration.OpenBraceToken.TrailingTrivia.All(static t => t.IsWhitespace()))
+                semiColon = semiColon.WithAppendedTrailingTrivia(namespaceDeclaration.OpenBraceToken.TrailingTrivia);
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/CodeFixes/ConvertNamespace/ConvertNamespaceTransform.cs:196]: general idea here is that we move the trivia on the `{` to the `;` (excluding just pure whitespace runs).  This has an added benefit of making sure that newlines around the `{` become the newlines of the `;` which feels intuitive/naturel in terms of the resultant namespace.

File: src/Analyzers/CSharp/CodeFixes/ConvertNamespace/ConvertNamespaceTransform.cs [hunk 2]
```diff
@@ -186,53 +186,56 @@ public static (SourceText text, TextSpan semicolonSpan) ConvertNamespaceDeclarat
...
+                semiColon,
                 namespaceDeclaration.Externs,
                 namespaceDeclaration.Usings,
                 namespaceDeclaration.Members);
 
-            var firstBodyToken = fileScopedNamespace.SemicolonToken.GetNextToken();
+            // Copy trivia from the close brace to the end of the file scoped namespace (which means after all of the members)
+            fileScopedNamespace = fileScopedNamespace
+                .WithAppendedTrailingTrivia(namespaceDeclaration.CloseBraceToken.LeadingTrivia)
+                .WithAppendedTrailingTrivia(namespaceDeclaration.CloseBraceToken.TrailingTrivia);
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/CodeFixes/ConvertNamespace/ConvertNamespaceTransform.cs:212]: we also never drop any trivia on the close-brace token.

File: src/Analyzers/CSharp/CodeFixes/ConvertNamespace/ConvertNamespaceTransform.cs [hunk 3]
```diff
@@ -186,53 +186,56 @@ public static (SourceText text, TextSpan semicolonSpan) ConvertNamespaceDeclarat
...
+                    { RawKind: (int)SyntaxKind.EndOfLineTrivia or (int)SyntaxKind.EndIfDirectiveTrivia or (int)SyntaxKind.EndRegionDirectiveTrivia },
+                    { RawKind: (int)SyntaxKind.EndOfLineTrivia } ])
+                {
+                    break;
+                }
+
+                // New namespace has excess newlines, remove the last one and try again.
+                fileScopedNamespace = fileScopedNamespace.WithTrailingTrivia(
+                    trailingTrivia.Take(trailingTrivia.Count - 1));
             }
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/CodeFixes/ConvertNamespace/ConvertNamespaceTransform.cs:239]: new logic is different, but seems to fit better with the before code.  First, we continuously trim down any multiple-new-lines we have.  We also remove the final new-line if hte original code didn't end with a newline either.

File: src/Analyzers/CSharp/Tests/ConvertNamespace/ConvertToFileScopedNamespaceAnalyzerTests.cs [hunk 4]
```diff
@@ -18,8 +18,6 @@ namespace Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.ConvertNamespace
 
     public class ConvertToFileScopedNamespaceAnalyzerTests
     {
-        #region Convert To File Scoped
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Tests/ConvertNamespace/ConvertToFileScopedNamespaceAnalyzerTests.cs:21]: pointless region given the file.

File: src/Analyzers/CSharp/Tests/ConvertNamespace/ConvertToFileScopedNamespaceAnalyzerTests.cs [hunk 5]
```diff
@@ -349,7 +347,6 @@ class C
 ",
                 FixedCode = @"
 namespace $$N;
-
 // comment
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Tests/ConvertNamespace/ConvertToFileScopedNamespaceAnalyzerTests.cs:350]: this feels like an improvement.  the original code effectively had the comment right underneath the namespace.  the new code preserves that.

File: src/Analyzers/CSharp/Tests/ConvertNamespace/ConvertToFileScopedNamespaceAnalyzerTests.cs [hunk 6]
```diff
@@ -787,8 +784,8 @@ public async Task TestConvertToFileScopedSingleLineNamespace1()
 [|namespace N|] { class C { } }
 ",
                 FixedCode = @"
-namespace $$N; 
-class C { } ",
+namespace $$N; class C { }
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Tests/ConvertNamespace/ConvertToFileScopedNamespaceAnalyzerTests.cs:787]: another improvement. the original code was on the same line as the namespace.  we preserve that.

File: src/Analyzers/CSharp/Tests/ConvertNamespace/ConvertToFileScopedNamespaceAnalyzerTests.cs [hunk 7]
```diff
@@ -808,8 +805,8 @@ public async Task TestConvertToFileScopedSingleLineNamespace2()
 ",
                 FixedCode = @"
 namespace $$N;
-
-class C { } ",
+class C { }
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Tests/ConvertNamespace/ConvertToFileScopedNamespaceAnalyzerTests.cs:808]: improvement.  the original class was directly underneath the namespace.  the new code preserves that.

File: src/EditorFeatures/CSharpTest/ConvertNamespace/ConvertNamespaceRefactoringTests.cs [hunk 8]
```diff
@@ -478,6 +477,7 @@ public async Task TextConvertToFileScopedWithCommentedOutContents()
 ",
                 FixedCode = @"
 namespace N;
+
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/ConvertNamespace/ConvertNamespaceRefactoringTests.cs:480]: improvment. the original comment was two lines below the namespace.  we preserve that.

**Review by davidwengier — APPROVED** [review]
Looks like this also fixes https://github.com/dotnet/roslyn/issues/63792 ?

**Pull request merged**